Mathlib/CategoryTheory/Sites/ConstantSheaf.lean

```diff
@@ -45,22 +66,176 @@ sheafification of the constant presheaf.
+
+/--
+A sheaf is constant if the counit of the constant sheaf adjunction applied to it is an isomorphism.
+-/
+class IsConstant (F : Sheaf J D) : Prop where
+  mem_essImage : F ∈ (constantSheaf J D).essImage
+
+lemma mem_essImage_of_isConstant (F : Sheaf J D) [IsConstant J F] :
+    F ∈ (constantSheaf J D).essImage :=
+  IsConstant.mem_essImage
+
+lemma isConstant_congr {F G : Sheaf J D} (i : F ≅ G) [IsConstant J F] : IsConstant J G where
+  mem_essImage := essImage.ofIso i F.mem_essImage_of_isConstant
+
+lemma isConstant_of_iso {F : Sheaf J D} {X : D} (i : F ≅ (constantSheaf J D).obj X) :
+    IsConstant J F := ⟨_, ⟨i.symm⟩⟩
+
+lemma isConstant_iff_mem_essImage {L : D ⥤ Sheaf J D} {T : C} (hT : IsTerminal T)
+    (adj : L ⊣ (sheafSections J D).obj ⟨T⟩)
+    (F : Sheaf J D) : IsConstant J F ↔ F ∈ L.essImage := by
+  rw [essImage_eq_of_natIso (adj.leftAdjointUniq (constantSheafAdj J D hT))]
+  exact ⟨fun ⟨h⟩ ↦ h, fun h ↦ ⟨h⟩⟩
+
+instance (F : Sheaf J D) [HasTerminal C]
+    [IsIso <| (constantSheafAdj J D terminalIsTerminal).counit.app F] : IsConstant J F where
+  mem_essImage := ⟨_, ⟨asIso <| (constantSheafAdj J D terminalIsTerminal).counit.app F⟩⟩
+
+instance [(constantSheaf J D).Faithful] [(constantSheaf J D).Full] (F : Sheaf J D)
+    [IsConstant J F] {T : C} (hT : IsTerminal T) :
+    IsIso ((constantSheafAdj J D hT).counit.app F) := by
+  rw [isIso_counit_app_iff_mem_essImage]
+  exact F.mem_essImage_of_isConstant
+
+/--
+If the constant sheaf functor is fully faithful, then a sheaf is constant if and only if the
+counit of the constant sheaf adjunction applied to it is an isomorphism.
+-/
+lemma isConstant_iff_isIso_counit_app [(constantSheaf J D).Faithful] [(constantSheaf J D).Full]
+    (F : Sheaf J D) {T : C} (hT : IsTerminal T) :
+      IsConstant J F ↔ (IsIso <| (constantSheafAdj J D hT).counit.app F) :=
+  ⟨fun _ ↦ inferInstance, fun _ ↦ ⟨_, ⟨asIso <| (constantSheafAdj J D hT).counit.app F⟩⟩⟩
+
+/--
+A variant of `isConstant_iff_isIso_counit_app` for a general left adjoint to evaluation at a
+terminal object.
+-/
+lemma isConstant_iff_isIso_counit_app'  {L : D ⥤ Sheaf J D} {T : C} (hT : IsTerminal T)
+    (adj : L ⊣ (sheafSections J D).obj ⟨T⟩)
+    [L.Faithful] [L.Full] (F : Sheaf J D) : IsConstant J F ↔ IsIso (adj.counit.app F) :=
+  (isConstant_iff_mem_essImage J hT adj F).trans (isIso_counit_app_iff_mem_essImage adj).symm
+
+end Sheaf
+
+section Equivalence
+variable {C' : Type*} [Category C'] (K : GrothendieckTopology C') [HasWeakSheafify K D]
+variable (G : C ⥤ C') [∀ (X : (C')ᵒᵖ), HasLimitsOfShape (StructuredArrow X G.op) D]
+  [G.IsDenseSubsite J K] {T : C} (hT : IsTerminal T) (hT' : IsTerminal (G.obj T))
+
+open IsDenseSubsite
+
+variable (D) in
+/--
+The constant sheaf functor commutes up to isomorphism the equivalence of sheaf categories induced
+by a dense subsite.
+-/
+noncomputable def equivCommuteConstant : let e : Sheaf J D ≌ Sheaf K D := sheafEquiv G J K D
+    constantSheaf J D ⋙ e.functor ≅ constantSheaf K D :=
```

Comment:
In the statement, it would be better to use `letI`, or replace e by `sheafEquiv ...`.

Reply:
I only used `let e := ...` because the name of `sheafEquiv` used to be so long :-) 
I'll change it to `sheafEquiv` in the statement.